src/Tools/Source/CompilerGeneratorTools/Source/BoundTreeGenerator/BoundNodeClassWriter.cs

```diff
@@ -1438,6 +1439,103 @@ private void WriteRewriter()
             }
         }
 
+        private void WriteNullabilityRewriter()
+        {
+            switch (_targetLang)
+            {
+                case TargetLanguage.VB:
+                    break;
+
+                case TargetLanguage.CSharp:
+                    {
+                        Blank();
+                        WriteLine("internal sealed partial class NullabilityRewriter : BoundTreeRewriter");
+                        Brace();
+
+                        var topLevelNullabilities = "_topLevelNullabilities";
+                        WriteLine($"private readonly ImmutableDictionary<BoundExpression, TypeSymbolWithAnnotations> {topLevelNullabilities};");
+
+                        Blank();
+                        WriteLine("public NullabilityRewriter(ImmutableDictionary<BoundExpression, TypeSymbolWithAnnotations> topLevelNullabilities)");
+                        Brace();
+                        WriteLine($"{topLevelNullabilities} = topLevelNullabilities;");
+                        Unbrace();
+
+                        foreach (var node in _tree.Types.OfType<Node>().Where(n => IsDerivedType("BoundExpression", n.Name)))
+                        {
+                            Blank();
+                            WriteLine(GetVisitFunctionDeclaration(node.Name, isOverride: true));
+                            Brace();
+                            bool hadField = false;
+                            foreach (Field field in AllNodeOrNodeListFields(node))
+                            {
+                                hadField = true;
+                                if (SkipInVisitor(field))
+                                {
+                                    WriteLine($"{field.Type} {ToCamelCase(field.Name)} = node.{field.Name};");
+                                }
+                                else
+                                {
+                                    WriteLine($"{field.Type} {ToCamelCase(field.Name)} = ({field.Type})this.Visit{(IsNodeList(field.Type) ? "List" : "")}(node.{field.Name});");
```

Comment:
I copied the code from the `VisitRewriter` method. It may not be necessary, but unless we want to update that as well I'll keep this as is.

---
In reply to: [252404229](https://github.com/dotnet/roslyn/pull/32899#discussion_r252404229) [](ancestors = 252404229)